### PR TITLE
Introduce snapshot tests for widgets

### PR DESCRIPTION
**Pull request by zoechi**
Just wanted to check if you are interested in such snapshot tests.
I that case I'd try to get more tests added for all kinds of edge cases.

For me writing tests is a good way to better understand the code.

### CODE REVIEW

**Comment by Philipp-M** [comment]
Wow, really nice :)

I think right now it's probably a little bit early to add much testing infrastructure, as I think the API (including widgets) will likely still change quite a bit and then (adjusting) tests may slow down while experimenting.

But I'm fine with merging this and would like to revisit this later in more detail, when the dust (experiments) settles.

Btw. I think especially the widget side will likely see quite a few changes (this is not the most polished area of code).
E.g. I'm still not sure how to do styling contexts (e.g. whether the widget contexts (such as `PaintCx`) there should hold more state transferred to sub widgets or not etc.). Right now there's a slightly hacky escape hatch with `PaintCx::override_style` to e.g. optionally style when hovering without overriding the child style (via `StyleOnHover`).
Or whether modifiers such as `view.padding()` should be new widgets (thinking about cases such as `view.padding(10).bg(Color::Red).padding(15)` which for my reading should be a red view with 10 padding and around the red background padding of 15).
I'm leaning towards refactoring the layout to the SwiftUI layout protocol currently (and it also fits with the current `v_stack` and `h_stack` views), as it's a simple protocol (one tree traversal) and I hear good things about SwiftUIs DX (JPC has a very similar protocol btw. I think the only real difference is when I'm getting it right, is that the parent doesn't propose the size for the child).

**Comment by zoechi** [comment]
I would love to dive into discussions about these topics but I think I first need to understand the existing code better.

What I am currently thinking about is a mix of how it works in Flutter and web. I imagine a dedicated style structs for every widget but also supporting rules like when a widget is inside another widget, or when it's an odd child index of the parent, or the first/last child, ...

What I also like from Flutter is dedicated layout widgets for border, padding, ... instead of supporting such styling on every widget individually. To make that work with styles that would require to allow styles to wrap widgets with other widgets, or widgets would need to wrap their content with border, padding, ... widgets and only apply passed styles, but that would be less flexible.


I wouldn't mind just throwing tests out if they turn out to cause too much work after certain changes, but I find them helpful for development, so I'm glad you are willing to merge them.

**Comment by Philipp-M** [comment]
I guess it depends on each widget, I kinda want to try to steer towards how SwiftUI (and JPC) works, as it was also one of the main inspiration for xilem (and thus trui, which is basically xilem for TUI). AFAIK it's not that far from Flutter in a lot of ways but from what I've heard and seen a little bit more polished.
But as said this is quite an experiment still, I want this (similar as xilem itself) to go to whatever works best for all kinds of use cases. I'd like to keep the necessity to write custom widgets low, so that most stuff can be modeled via simple views (and view modifiers as SwiftUI and JPC calls them) directly. I think the button example kinda hints that direction already slightly.

Styling such widgets built via view composition could e.g. then be done via structs given to functions (`BorderStyle`  could probably be seen as example, I've experimented here already with a little bit of syntax sugar `impl Into<BorderStyle>`)

**Comment by Philipp-M** [comment]
Well I started answering, but I instead put it into a separate issue where it probably makes more sense (I think the issue content kinda fits parts of what you mean?)

**Comment by Philipp-M** [comment]
> I imagine a dedicated style structs for every widget but also supporting rules like when a widget is inside another widget, or when it's an odd child index of the parent, or the first/last child

Btw. good input, it would be nice to reach this as programmatically as possible (as for one I think evaluating the view tree should be rather cheap, and because the user has more control over it, e.g. weird cases such as a fibonacci background coloring for list items in a list or whatever crazy ideas the user has^^).
At the same time this should optimally be a reusable view modifier, to compose with other views/viewsequences.
Could e.g. be done via an iterator (or whatever type wizardry is necessary).

**Comment by zoechi** [comment]
I imagine having a set of rules like first-of, last-of, even, odd, ..., but also allowing to also pass custom rules as trait object or closure

**Comment by zoechi** [comment]
I added another helper to test a view because I wasn't able to easily create widgets withouts views in all cases.
This was mostly an exercise in learing about the existing code and of course making further development with tests easier.

File: src/app.rs [hunk 1]
```diff
@@ -392,6 +418,24 @@ impl<T: Send + 'static, V: View<T> + 'static> App<T, V> {
...
+        &mut self.terminal
+    }
+}
+
+/// Restore the terminal no matter how the app exits
+impl<T: Send + 'static, V: View<T> + 'static> Drop for App<T, V> {
+    fn drop(&mut self) {
+        self.cleanup()
+            .unwrap_or_else(|e| eprint!("Restoring the terminal failed: {e}"));
+    }
```
Comment (Philipp-M) [inline src/app.rs]: Nice :)

File: src/test_helper.rs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+/// Render a view and return the terminal to check the generated output
+///
+/// * `buffer_size` - The terminal output buffer is set to that size.
+/// * `sut` - A closure that returns the widget to test when called.
+///   This expects a function so it can create it inside the thread of the [`App`].
+/// * `state` - Is the state for the [`App`]
+pub fn render_view<T: Send + 'static>(
+    buffer_size: Size,
+    sut: Box<dyn Fn() -> Box<dyn AnyView<T>> + Send + Sync>,
```
Comment (Philipp-M) [inline src/test_helper.rs]: It should be possible to just use an `Arc<impl View>` instead, similar as in the blocks example. (Not that it makes a massive difference though)

File: src/test_helper.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        phantom: PhantomData,
+    }
+}
+
+impl<T, A, V> ViewMarker for DebugView<V, T, A> {}
+
+impl<T, A, V> View<T, A> for DebugView<V, T, A>
+where
+    V: View<T, A>,
+    V::Element: 'static,
```
Comment (Philipp-M) [inline src/test_helper.rs]: Not necessary `Widget` has this already as super bound

File: src/test_helper.rs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+    fn build(&self, cx: &mut Cx) -> (xilem_core::Id, Self::State, Self::Element) {
+        let (id, (state, element)) = cx.with_new_id(|cx| {
+            let (child_id, state, element) = self.content.build(cx);
+            (
+                (state, child_id),
+                DebugWidget::new(element, self.debug_chan_tx.clone()),
+            )
+        });
+        (id, state, element)
```
Comment (Philipp-M) [inline src/test_helper.rs]: Since this view doesn't take any messages directly, it's probably simpler to not use a separate id at all (needs some adjustments in `View::State` and the other methods as well)
```suggestion
        let (id, state, element) = self.content.build(cx);
        (id, state, DebugWidget::new(element, self.debug_chan_tx.clone()))
```

File: src/test_helper.rs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        element: &mut Self::Element,
+    ) -> crate::widget::ChangeFlags {
+        let changeflags = ChangeFlags::empty();
+
+        let element = element
+            .content
+            .downcast_mut()
+            .expect("The DebugView content widget changed its type, this should never happen!");
+
+        changeflags
```
Comment (Philipp-M) [inline src/test_helper.rs]: I think changeflags can just be omitted, they're not doing anything here

File: src/test_helper.rs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    ) -> MessageResult<A> {
+        if let Some((first, rest_path)) = id_path.split_first() {
+            if first == child_id {
+                self.content.message(rest_path, state, message, app_state)
+            } else {
+                xilem_core::MessageResult::Stale(message)
+            }
+        } else {
+            xilem_core::MessageResult::Nop
+        }
```
Comment (Philipp-M) [inline src/test_helper.rs]: This can just be 
```suggestion
        self.content.message(id_path, state, message, app_state)
```
then

File: src/widget/block.rs [hunk 7]
```diff
@@ -209,3 +209,38 @@ impl Widget for Block {
...
+
+    use super::*;
+
+    #[test]
+    fn simple_text_test() {
+        let content = Text {
+            text: "some text".into(),
+            style: Style::default().fg(Color::Red),
+        };
+        let mut sut = Block::new(
```
Comment (Philipp-M) [inline src/widget/block.rs:229]: Just out of interest what does `sut` mean (as it's not immediately clear)?

File: src/widget/core.rs [hunk 8]
```diff
@@ -28,6 +37,7 @@ pub enum Event {
 }
 
 /// Static state that is shared between most contexts.
+#[derive(Debug)]
```
Comment (Philipp-M) [inline src/widget/core.rs]: Not sure where this is necessary but with it, it doesn't compile since `Message` needs to be `Debug` (and I'm not sure whether we want this to be `Debug` anyway)

File: src/widget/block.rs (hunk 7 above)
Comment (zoechi) [reply src/widget/block.rs:229]: "system under test" I thought it was common enugh to just use it.
This way the central subject stands out a bit in a test.

File: src/widget/core.rs (hunk 8 above)
Comment (zoechi) [reply src/widget/core.rs]: I thought about https://rust-lang.github.io/api-guidelines/interoperability.html#c-common-traits and because I wanted it for debugging myself I added it and left it.
If you don't intend to go in that direction I'll just remove it and also stay away from adding it to others.

**Comment by zoechi** [comment]
Thanks a lot for the feedback!

File: src/widget/core.rs (hunk 8 above)
Comment (Philipp-M) [reply src/widget/core.rs]: It's just that I think that there'll likely be (most) stuff that's not implementing `Debug` in there anyway (but of course we could do a manual impl, filtering the printable/serializable stuff).
`Message` is already an example (although of course we could do a manual impl as well there, likely only printing the `IdPath`). Not sure...

I don't think `CxState` is a real public type though (the wrappers like `EventCx` etc. are).

But right now it doesn't compile with this...

Comment (zoechi) [reply src/widget/core.rs]: I couldn't reproduce the build issue, but I'll just remove it.

Comment (Philipp-M) [reply src/widget/core.rs]: Ok that's interesting. For me this is coming:

```
error[E0277]: `Message` doesn't implement `Debug`
  --> src/widget/core.rs:42:5
   |
40 | #[derive(Debug)]
   |          ----- in this derive macro expansion
41 | pub struct CxState<'a> {
42 |     messages: &'a mut Vec<Message>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Message` cannot be formatted using `{:?}`
```

File: src/test_helper.rs (hunk 2 above)
Comment (zoechi) [reply src/test_helper.rs]: That's much better

**Comment by zoechi** [comment]
All suggestions should be addressed

**Comment by zoechi** [comment]
All suggestions should be addressed

File: src/app.rs (hunk 1 above)
Comment (Philipp-M) [reply src/app.rs]: One thing to keep in mind:

Currently the real terminal/stdout is mutated in tests when I see it right (`EnableFocusChange`, `EnableMouseCapture`...). It would be good to redirect or disable these mutations.


Hmm well thinking about it, a `#[cfg(not(test))]` above each of the `execute!(stdout(),..)` should be enough already, so maybe add this as well. And I think above `enable/disable_raw_mode` as well.

File: src/test_helper.rs [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+    fn rebuild(
+        &self,
+        _cx: &mut Cx,
+        _prev: &Self,
+        _id: &mut xilem_core::Id,
+        _state: &mut Self::State,
+        _element: &mut Self::Element,
+    ) -> crate::widget::ChangeFlags {
+        ChangeFlags::empty()
```
Comment (Philipp-M) [inline src/test_helper.rs]: Well the child view should still be rebuilt, otherwise no updates can be done to this view, I just meant that the extra variable was unnecessary.

```suggestion
        cx: &mut Cx,
        prev: &Self,
        id: &mut xilem_core::Id,
        state: &mut Self::State,
        element: &mut Self::Element,
    ) -> ChangeFlags {
        let element = element
            .content
            .downcast_mut()
            .expect("The DebugView content widget changed its type, this should never happen!");
        self.content.rebuild(cx, &prev.content, id, state, element)
```

**Review by Philipp-M — COMMENTED** [review]
I think when the two comments are addressed, it's good to merge.

File: src/app.rs (hunk 1 above)
Comment (zoechi) [reply src/app.rs]: This shouldn't be necessary. Tests use `Terminal<TestBackend>` for widget and view tests.

> /// A [`Backend`] implementation used for integration testing that that renders to an in memory
/// buffer.

Do you think these commands by itself already alter stdout?
Perhaps if snapshot tests run in parallel with normal tests?

I didn't think it worth it to clutter the test with `[cfg(not(test))]` but I wouldn't mind changing it if you think it's worth it.

Comment (Philipp-M) [reply src/app.rs]: Sure, but doesn't `execute!(stdout(),...)` directly act on stdout/terminal (i.e. circumvent ratatui)?

Comment (zoechi) [reply src/app.rs]: Ok, I fixed it.

File: src/app.rs [hunk 10]
```diff
@@ -392,6 +437,25 @@ impl<T: Send + 'static, V: View<T> + 'static> App<T, V> {
...
         Ok(())
     }
+
+    #[cfg(test)]
+    pub fn get_event_tx(&self) -> tokio::sync::mpsc::Sender<Event> {
+        self.event_tx.clone()
+    }
+
+    #[cfg(test)]
+    pub fn get_terminal(&mut self) -> &mut Terminal<TestBackend> {
```
Comment (Philipp-M) [inline src/app.rs]: I'd like to follow the [API guidelines of Rust](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter)
```suggestion
    pub fn terminal_mut(&mut self) -> &mut Terminal<TestBackend> {
```

File: src/app.rs [hunk 11]
```diff
@@ -287,11 +312,19 @@ impl<T: Send + 'static, V: View<T> + 'static> App<T, V> {
             };
 
             root_pod.paint(&mut paint_cx);
-            execute!(stdout(), BeginSynchronizedUpdate)?;
+
+            #[cfg(not(test))]
+            {
+                execute!(stdout(), BeginSynchronizedUpdate)?;
+                self.terminal.flush()?;
```
Comment (Philipp-M) [inline src/app.rs]: Shouldn't this still be enabled? (as well as the flush swap_buffers and flush below)?

(So just above the `execute!` macros)

**Review by Philipp-M — COMMENTED** [review]
Sorry to be a little bit nitpicky...

File: src/app.rs (hunk 10 above)
Comment (zoechi) [reply src/app.rs]: Not at all. I appreciate any feedback. I misremembered the get_ prefix and wasn't aware of the _mut suffix at all.

File: src/app.rs (hunk 11 above)
Comment (zoechi) [reply src/app.rs]: I somehow missed the swap_buffers(), but the flush() is covered in line 327.

Comment (Philipp-M) [reply src/app.rs]: You're right, missed that.

I pushed this and some smaller fixes to get it ready to merge, hope that's fine for you.

**Review by Philipp-M — APPROVED** [review]
I did a last push for smaller fixes not worth discussing (IMO), hope that's fine for you.

Otherwise LGTM, thanks :)
